### ISSUE TITLE
Theoretical Question Regarding Target Variable Encoding

### ISSUE THREAD

**Issue by leoanghelescu** [Intel-bigdata/imllib-spark#16]
I am stepping through the code to better understand the FFM calculation and the Stochastic Gradient calculation. Prior to this, I notice in the code of FFMExample.scala, when we input the data:

//LA: split by spaces and parse out the 0, then drop ...
val data= sc.textFile(args_manual(0)).map(_.split("\\s")).map(x => {
  //LA: parse the first element, if 1, then map to value 1, else map to -1
  val y = if(x(0).toInt > 0 ) 1.0 else -1.0
  //LA: the rest of the elements are parsed out and split by
  val nodeArray: Array[(Int, Int, Double)] = x.drop(1).map(_.split(":")).map(x => {
    (x(0).toInt, x(1).toInt, x(2).toDouble)
  })
  (y, nodeArray) //end up with y (+/- 1)
}).repartition(4)

We assign a Y value of -1 to the labels if they are zero. I am having a hard time to understand why this is. May you please give me a brief explanation?

Thanks!
Leo

**Comment by VinceShieh** [comment]
Hi, it is more of an engineering consideration. You can check the update formula for FFM, such as in "-label * expnyt / (1 + expnyt)", basically zero carries less information.
It's not an FFM specific but a usual implementation method easy to be found in other models such as SVM, LR, etc..

**Comment by leoanghelescu** [comment]
Thanks Vincent, makes perfect sense.

Question at the end about the final transformation (I'm expecting a logit function)  of the predict return value:

In FFMExample.scala

    val scores: RDD[(Double, Double)] = testing.map(x => {
      val p = ffm.predict(x._2)
      val ret = if (p >= 0.5) 1.0 else -1.0
      (ret, x._1)
    })

And the returned value of predict is incremented as such and never transformed (to a 0-1 range)

 t += weights(w1_index + d) * weights(w2_index + d) * v

But i do see this logit trhansfor when we are doing computeFFM

    val expnyt = math.exp(-label * t)
    val tr_loss = math.log(1 + expnyt)

My main question really is, how do we know that the return value from predict it range (0,1)?

Thanks again for your help!!
Leo

**Comment by VinceShieh** [comment]
duplicated issue.
